keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,150 @@
+---
+title: Skip Volume Ownership Change
+authors:
+  - "@gnuified"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-auth
+reviewers:
+  - "@msau42"
+  - "@liggit"
+  - "@tallclair"
+approvers:
+  - "@saad-ali"
+editor: TBD
+creation-date: 2020-01-20
+last-updated: 2020-01-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Skip Volume Ownership Change
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `pod.Spec.SecurityContext.FSGroupPermissionChangePolicy` and
```

Comment:
Should this be a pod-level setting, or a per-volume setting?

Reply:
I moved this field to `pod.spec.volume`. I think will work better there and allow an user to have more granular control over volume permission change.

Reply:
I'm actually not sure how much this helps? fsgroup is set per pod and moving it to `pod.spec.volume` is still per pod...

Reply:
I am thinking that - it would allow possibly to have different policies for different volumes in a pod.  I know today that is not the case, but it makes sense to futureproof api little bit.

Reply:
It may cause less confusion for now to keep them together in the same place, and then move both of them out in the future.

Reply:
hmm, I am not sure what could be confusing about having the field in `pod.Spec.Volumes[]` . Is that purely because `fsGroup` is present in `pod.spec.SecurityContext` ? There are other things I am thinking - such that if you apply a policy to a volume type which can't support it, we have the opportunity to reject such change, but that again depends on having more volume-type specific policies.

I do not have very strong opinion on this, but I am right now leaning towards keeping them in `pod.Spec.Volumes[]`. @tallclair what are your thoughts on this?

Reply:
I guess you could have the same fsgroup across all volumes, but different policies per volume. However, this also makes it harder if there is an admission controller/hook that is injecting volumes into Pods.

Reply:
So I guess the confusing part for such an automated tool is  - it will not know for which volume types it should set `permisionchangepolicy`?

Reply:
I could not see these comments in review page?

So I actually bring up the same question. I kind of prefer pod level in pod.Spec.SecurityContext. 
1. It is easy for user to set fsGroup and fsGoupChangePolicy together.
2. This policy is applied to all volumes except for atomic volumes. as @msau42 mentioned the performance should not be an issue for atomic volumes.

I feel there is no strong reason to set volume level.

Reply:
alright. sure. I will move it.